### PR TITLE
Detect multiprotocol keywords 7304 v2

### PR DESCRIPTION
**Pull request by catenacyber**
Link to ticket: https://redmine.openinfosecfoundation.org/issues/
https://redmine.openinfosecfoundation.org/issues/7304

Describe changes:
- cleaner code for multi app-layer keywords such as ja4

Is there something to do with DCERPC/SMB stuff ? cc @inashivb 

I think file keywords have their own logic and it is ok.

#12149 with better commit messages

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 23713

File: src/detect-parse.c [hunk 1]
```diff
@@ -1735,6 +1735,84 @@ int DetectSignatureAddTransform(Signature *s, int transform, void *options)
...
+                break;
+            }
+            // first disable the ones that do not match
+            bool found = false;
+            while (*alprotos != ALPROTO_UNKNOWN) {
+                if (s->init_data->alprotos[i] == *alprotos) {
+                    found = true;
+                    break;
+                }
+                alprotos++;
```
Comment (inashivb) [inline src/detect-parse.c:1766]: This looks like we'll exhaust alprotos on checks against the first alproto entry of the signature if there was no match. Is this the intention?

**Comment by inashivb** [comment]
> Is there something to do with DCERPC/SMB stuff ?

As far as I can tell, DCERPC/SMB will always fall in the default `DetectSignatureSetAppProto` case as there's just one to one protocol compatibility. So, nothing to do specially for that, I think.

File: src/detect-parse.c (hunk 1 above)
Comment (catenacyber) [reply src/detect-parse.c:1766]: Not sure I follow what you mean here...

There is a double loop with the current signature alprotos, and the new ones...

**Comment by catenacyber** [comment]
> > Is there something to do with DCERPC/SMB stuff ?
> 
> As far as I can tell, DCERPC/SMB will always fall in the default `DetectSignatureSetAppProto` case as there's just one to one protocol compatibility. So, nothing to do specially for that, I think.

DCERPC/SMB is not totally default as can be seen in `AppProtoCommon` :

> // ok to have dcerpc keywords in smb sig

My question is like Could we get rid of this hardcoded behavior and have dcerpc keywords just call `DetectSignatureSetMultiAppProto` with both DCERPC and SMB ? (does it make sense, etc...)

> Another thought: why do we not just have a protocol compatibility matrix and look through it if a protocol is compatible with another one and what should be the final signature protocol in that case?

Because of ja3/ja4 which are both quic and tls... (files keywords also, etc...)

File: src/detect-parse.c (hunk 1 above)
Comment (inashivb) [reply src/detect-parse.c:1766]: I mean for a say, `s->init_data->alprotos = [proto1, proto2, proto3, ALPROTO_UNKNOWN]`
and `alprotos = [proto4, proto3, ALPROTO_UNKNOWN]` coming to this part of the code,

First pass of the for loop looking for `proto1` will bring `alprotos` pointer out of the array. We won't go over `proto2` and `proto3` in the `s->init_data->alprotos`, is this possible/intended?

Comment (catenacyber) [reply src/detect-parse.c:1766]: Very right

**Comment by inashivb** [comment]
> > > Is there something to do with DCERPC/SMB stuff ?
> > 
> > 
> > As far as I can tell, DCERPC/SMB will always fall in the default `DetectSignatureSetAppProto` case as there's just one to one protocol compatibility. So, nothing to do specially for that, I think.
> 
> DCERPC/SMB is not totally default as can be seen in `AppProtoCommon` :

I think I'm misunderstanding what you mean by "default" here. Could you please explain?
I'm purely talking from the perspective of smb rules allowing only dcerpc keywords hence there not being a need for "multiple" protocol compatibility checks.

> 
> > // ok to have dcerpc keywords in smb sig
> 
> My question is like Could we get rid of this hardcoded behavior and have dcerpc keywords just call `DetectSignatureSetMultiAppProto` with both DCERPC and SMB ? (does it make sense, etc...)
> 

yes, you could just call `DetectSignatureSetMultiAppProto` and it does make sense.
What I was saying was that even if you do call `DetectSignatureSetMultiAppProto` with DCERPC and SMB, it will fall into https://github.com/OISF/suricata/pull/12229/files#diff-38b98a1ee88485d56b7f5a35a5cea0bf27481910a7cf16ef4dc58b782ea9ccc6R1787 block of code ultimately calling 
`AppProtoCommon` for this case since there is just one intersection possible.
Is this an incorrect assessment?

File: src/detect-parse.c (hunk 1 above)
Comment (catenacyber) [reply src/detect-parse.c:1766]: Thanks for spotting this

**Comment by catenacyber** [comment]
DCERPC/SMB has not a default behavior where default means not a specific case in the `switch` for `AppProtoEquals` and `AppProtoCommon`

Looking into the code to understand what is going on, I find some problems :

1. Why can we write the first rule and not the second one ?
```
alert tcp any any -> any any (flow:to_server; smb.version: 2; dcerpc.iface:367abb81-9844-35f1-ad32-98f038001003; dcerpc.opnum:15; sid:1;)
alert tcp any any -> any any (flow:to_server; dcerpc.iface:367abb81-9844-35f1-ad32-98f038001003; dcerpc.opnum:15; smb.version: 2; sid:1;)
```
Just having the smb keyword first or after dcerpc one...

2. How can we express that we want a rule to match some dceprc keywords, but only over SMB ? or only over raw non-SMB DCERPC ?

Looking at rules from dcerpc-smb-test-01 which are the same except their beginning : `alert tcp` `alert smb` and `alert dcerpc`...

So is it correct that `DetectSignatureSetMultiAppProto` could solve these problems ? (are they real problems)

Also wondering why I created ALPROTO_DOH2 when DCERPC over SMB, is kind of the same as DNS over HTTP2...
Maybe I should have rather implemented some `AppLayerRegisterMayContain(ALPROTO_HTTP2, ALPROTO_DNS)`

**Comment by catenacyber** [comment]
Next in https://github.com/OISF/suricata/pull/12310

**Pull request closed without merging**